exercises/complex-numbers/test/test_complex_numbers.c

```diff
@@ -0,0 +1,192 @@
+#include "../src/complex_numbers.h"
+
+#include "vendor/unity.h"
+
+#include <math.h>
+
+#define PI acos(-1)
+#define E exp(1)
+
+// imaginary unit
+void test_imaginary_unit(void) {
+    Complex z1;
+    z1.real = 0.0;
+    z1.imag = 1.0;
+
+    Complex z2;
+    z2.real = 0.0;
+    z2.imag = 1.0;
+
+    Complex result = c_mul(&z1, &z2);
+
+    TEST_ASSERT_EQUAL(result.real, -1.0);
+    TEST_ASSERT_EQUAL(result.imag, 0.0);
+}
+
+
+// addition
+void test_addition(void) {
+    Complex cases[][3] = {
+        { { .real = 1.0, .imag = 0.0 }, { .real = 2.0, .imag = 0.0 }, { .real = 3.0, .imag = 0.0 } },
+        { { .real = 0.0, .imag = 1.0 }, { .real = 0.0, .imag = 2.0 }, { .real = 0.0, .imag = 3.0 } },
+        { { .real = 1.0, .imag = 2.0 }, { .real = 3.0, .imag = 4.0 }, { .real = 4.0, .imag = 6.0 } }
+    };
+
+    unsigned int i;
+    for (i = 0; i < sizeof(cases) / sizeof(Complex[3]); i++) {
+        Complex result = c_add(&cases[i][0], &cases[i][1]);
+        
+        TEST_ASSERT_EQUAL(result.real, cases[i][2].real);
+        TEST_ASSERT_EQUAL(result.imag, cases[i][2].imag);
+    }
```

Comment:
@wolf99 @patricksjackson what are your thoughts on using loops like this here? It definitely makes things more compact, but it 1) loses the meaningful names for the sub-tests (ie: add only real, add only imaginary), 2) makes it more difficult for someone to figure out which sub-test is failing, and 3) isn't really consistent with how the other test suites are written.

Reply:
I'm fine with it as long as it's clear what's getting tested while reading this source, and that the error messages are clear on failure.

Here's my suggestion. I'd check the following two methods to see which gives the better error messages

```
void compare_complex(complex_t lhs, complex_t rhs) {
  TEST_ASSERT_EQUAL_DOUBLE(lhs.real, rhs.real);
  TEST_ASSERT_EQUAL_DOUBLE(lhs.imag, rhs.imag);
}

void test_addition(void) {
  compare_complex(c_add({.real = 1.0, .imag = 0.0}, {.real = 2.0, .imag = 0.0}), {.real = 3.0, .imag = 0.0});
  // etc
}
```
Or
```
bool compare_complex(complex_t lhs, complex_t rhs) {
  return (lhs.real == rhs.real) && (lhs.imag == rhs.imag);
}
void test_addition(void) {
  TEST_ASSERT(compare_complex(c_add({.real = 1.0, .imag = 0.0}, {.real = 2.0, .imag = 0.0}), {.real = 3.0, .imag = 0.0}));
  // etc
}
```

Reply:
How about I just make a separate function for each test case? It's not a big deal to do that.